File: LeetCode/46全排列.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    LinkedList<Integer> path = new LinkedList<>();

    public List<List<Integer>> permute(int[] nums) {
        boolean[] used = new boolean[nums.length];
        Arrays.fill(used, false);
        backtracking(nums, used);
        return ans;
    }

    public void backtracking(int[] nums, boolean[] used) {
        if (path.size() == nums.length) {
            ans.add(new ArrayList<>(path));
            return ;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            used[i] = true;
            path.add(nums[i]);
            backtracking(nums, used);
            path.removeLast();
            used[i] = false;
        }
    }
}

File: LeetCode/46全排列.go
var ans [][]int
var path []int
var used []bool

func permute(nums []int) [][]int {
	ans = [][]int{}
	path = []int{}
	used = make([]bool, len(nums))
	backtracking(nums)
	return ans
}

func backtracking(nums []int) {
	if len(path) == len(nums) {
		temp := make([]int, len(nums))
		copy(temp, path)
		ans = append(ans, temp)
	}
	for i := 0; i < len(nums); i++ {
		if used[i] == true {
			continue
		}
		used[i] = true
		path = append(path, nums[i])
		backtracking(nums)
		path = path[:len(path)-1]
		used[i] = false
	}
}